src/Raven.Client/Documents/InMemoryDocumentSessionOperations.cs

```diff
@@ -49,6 +50,27 @@ public abstract class InMemoryDocumentSessionOperations : IDisposable
         private BatchOptions _saveChangesOptions;
 
         /// <summary>
+        /// Store events
+        /// </summary>
+        public delegate void BeforeStoreDelegate(string id, object entityInstance, IDictionary<string, string> metadata, InMemoryDocumentSessionOperations session);
+        public event BeforeStoreDelegate BeforeStoreEvent;
```

Comment:
We should be able to register to those events either at the session level or at the store level